### DISCUSSION TITLE
About the development of modules

### DISCUSSION THREAD

**Discussion by buffcow** [LSPosed/LSPosed#1528] [Q&A] [answered]
Hello devs, I have a problem now:
I am writing a module to hook the system ui (com.android.systemui). There is a class in my module that extends a certain class of the system UI, but after loading the module, it will throw an error that this class cannot be found. What should I do how to extends a classe from the hooked application in a module?
![IMG_20220114_003831](https://user-images.githubusercontent.com/42611305/149371472-28f2b29a-2612-4f17-b304-9f363b4fa8db.jpg)
![IMG_20220114_004019](https://user-images.githubusercontent.com/42611305/149371511-0fcfe1b2-2a7b-45ce-b21b-f46a2a5deb1f.jpg)

CN: 我在写一个模块hook系统ui(com.android.systemui)，我的模块里有一个类继承了系统UI的某一个类，但是加载模块后会抛出找不到这个类的错误，请问我应该怎样在模块里继承被hook的应用里的类？

**Comment by buffcow** [comment]
@yujincheng08 帮我看一下呗🥺🙏

**Comment by ghost** [comment]
You should hook the class and not inject a new class. If you extend "a certain class of the system" how should the system use the new class? Only with injection of new class, but xposed/lsposed don't inject class but hook methods.
I think you should use a different approach.
Hook the methods which you override and replace with your code.

**Comment by yujincheng08** [comment, marked as answer]
模块的类加载器不是注入应用的类加载器，所以你继承注入应用的类提示找不到父类很正常。如果硬要继承，那就在该子类初始化前（第一次使用前）替换其类加载器的父加载器为注入应用的类加载器。

**Reply by anonymix007** [reply]
How to do that with the modern api?
I've tried the following without success:
```java
Field parent = ClassLoader.class.getDeclaredField("parent");
parent.setAccessible(true);
parent.set(getClassLoader(), param.getClassLoader());
```
